extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,147 @@
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.base.verify.Require;
+import io.deephaven.chunk.attributes.Any;
+import io.deephaven.engine.page.ChunkPage;
+import io.deephaven.parquet.table.pagestore.topage.ToPage;
+import io.deephaven.parquet.base.ColumnChunkReader;
+import io.deephaven.parquet.base.ColumnPageReader;
+import org.apache.parquet.internal.column.columnindex.OffsetIndex;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+final class OffsetIndexBasedColumnChunkPageStore<ATTR extends Any> extends ColumnChunkPageStore<ATTR> {
+    private static final long PAGE_SIZE_NOT_FIXED = -1;
+
+    private final OffsetIndex offsetIndex;
+    private final int numPages;
+    /**
+     * Fixed number of rows per page. Set as positive value if first ({@link #numPages}-1) pages have equal number of
+     * rows, else equal to {@value #PAGE_SIZE_NOT_FIXED}. We don't care about the size of the last page, because we can
+     * assume all pages to be of the same size and calculate the page number as
+     * {@code row_index / fixed_page_size -> page_number}. If it is greater than {@link #numPages}, we can assume the
+     * row to be coming from last page.
+     */
+    private final long fixedPageSize;
+
+    private final class PageState {
+        WeakReference<PageCache.IntrusivePage<ATTR>> pageRef;
+
+        PageState() {
+            // Initialized when used for the first time
+            pageRef = null;
+        }
+    }
+
+    private final AtomicReferenceArray<PageState> pageStates;
+    private final ColumnChunkReader.ColumnPageDirectAccessor columnPageDirectAccessor;
+
+    OffsetIndexBasedColumnChunkPageStore(@NotNull final PageCache<ATTR> pageCache,
+            @NotNull final ColumnChunkReader columnChunkReader,
+            final long mask,
+            @NotNull final ToPage<ATTR, ?> toPage) throws IOException {
+        super(pageCache, columnChunkReader, mask, toPage);
+        offsetIndex = columnChunkReader.getOffsetIndex();
+        Assert.assertion(offsetIndex != null, "offsetIndex != null");
+        numPages = offsetIndex.getPageCount();
+        Assert.assertion(numPages > 0, "numPages > 0");
+        pageStates = new AtomicReferenceArray<>(numPages);
+        columnPageDirectAccessor = columnChunkReader.getPageAccessor();
+
+        if (numPages == 1) {
+            fixedPageSize = numRows();
+            return;
+        }
+        boolean isPageSizeFixed = true;
+        final long firstPageSize = offsetIndex.getFirstRowIndex(1) - offsetIndex.getFirstRowIndex(0);
+        for (int i = 2; i < numPages && isPageSizeFixed; ++i) {
+            if (offsetIndex.getFirstRowIndex(i) - offsetIndex.getFirstRowIndex(i - 1) != firstPageSize) {
+                isPageSizeFixed = false;
+            }
```

Comment:
A more typical pattern IMO is to not have the boolean checked in the for loop, but just use a `break` here after setting `isPageSizeFixed = false`.